### DISCUSSION TITLE
How to configure auth via Secrets

### DISCUSSION THREAD

**Discussion by SimonSchmid** [nats-io/k8s#783] [Q&A]
Hi!

We just installed the 1.0.0 nats helm chart to our K8s cluster and want to set up auth. We are not sure how we achieve best what we want to do and I wanted to ask for some ideas or best practices.

We will mainly use the JetStream functionality and have multiple services that will connect to nats. Each service should have its own account so that we can set permissions accordingly as not all services should have sub and pub access to all streams and subjects.

We were thinking of having a Secret for each service that would contain username and password and could be mounted by the service. Important would be that the password is randomly generated and not specified statically in some helm chart. We struggle though as we are not sure how we would integrate those credentials into the auth configuration that nats expects. A solution we imagine could be to have the nats auth config as a template (defining users and permissions) and a job/controller running that reads in the different secrets to fill the passwords into the config template. The resulting config could then be mounted as configmap into nats. 

Does that approach make any sense? Or are we maybe overlooking some easier solution?

Best and thanks in advance,
Simon

**Comment by Parcley** [comment]
I know it has been a while, but afaik the basic questions are still unanswered:

- randomly creating the initial secrets at deployment time
- storing them in Kubernetes secrets at deployment time
- using the Kubernetes secrets in the nats helm chart (although I can see partly solutions to this point, e.g. : https://artifacthub.io/packages/helm/nats/nats#nats-config-includes )

Did you solve your issue in the mean time?
Maybe someone can jump in and help us out

**Comment by SimonSchmid** [comment]
We solved that by using the Auth Callout feature that was released in the meantime: https://docs.nats.io/running-a-nats-service/configuration/securing_nats/auth_callout. That requires an own implementation of a an auth service but allows full flexibility, so that we are even able to use existing service accounts in our system.
